src/core/base.jl

```diff
@@ -1,85 +1,50 @@
-    pm, gm = build_generic_model(power_data, gas_data, power_model_constructor, gas_model_constructor, post_method; power_ref_extensions=power_ref_extensions, kwargs...)
-    solution = solve_generic_model(pm, gm, solver; solution_builder = solution_builder)
-    return solution
-end
-
 ""
-function build_generic_model(pfile::String, gfile::String, power_model_constructor, gas_model_constructor, post_method; kwargs...)
-    gas_data = GasModels.parse_file(gfile)
-    power_data = PowerModels.parse_file(pfile)
-    return build_generic_model(power_data, gas_data, power_model_constructor, gas_model_constructor, post_method; kwargs...)
-end
-
-" Dummy post method to do nothing for the individual system"
-function empty_post_method(m; kwargs...)
+function instantiate_model(gfile::String, pfile::String, gtype::Type, ptype::Type, build_method; gext=[], pext=[], kwargs...)
+    gdata, pdata = [_GM.parse_file(gfile), _PM.parse_file(pfile)]
+    return instantiate_model(gdata, pdata, gtype, ptype, build_method; gext=gext, pext=pext, kwargs...)
 end
 
 ""
-function build_generic_model(pdata::Dict{String,Any}, gdata::Dict{String,Any}, power_model_constructor, gas_model_constructor, post_method; power_ref_extensions=[], multinetwork=false, multiconductor=false, kwargs...)
-    gm = GasModels.build_generic_model(gdata, gas_model_constructor, empty_post_method; multinetwork=multinetwork, kwargs...)
-    pm = PowerModels.build_model(pdata, power_model_constructor, empty_post_method; ref_extensions=power_ref_extensions, multinetwork=multinetwork, multiconductor=multiconductor)
+function instantiate_model(gdata::Dict{String,<:Any}, pdata::Dict{String,<:Any}, gtype::Type, ptype::Type, build_method; gext=[], pext=[], kwargs...)
+    # Instantiate the separate (empty) infrastructure models.
+    gm = _GM.instantiate_model(gdata, gtype, m->nothing; ref_extensions=gext)
+    pm = _PM.instantiate_model(pdata, ptype, m->nothing; ref_extensions=pext)
 
     add_junction_generators(pm, gm)
 
-    # a bit of a hack for now
+    # TODO: The below is a bit of a hack.
     gas_grid_per_unit(gm.data, pm.data)
 
-    # unify all the optimization models... a little bit of a hack...
+    # Unify all the optimization models.
     pm.model = gm.model
 
-    post_method(pm, gm; kwargs...)
-    return pm, gm
+    # Build the corresponding problem.
+    build_method(pm, gm, kwargs=kwargs)
+
+    return gm, pm
 end
 
 ""
-function solve_generic_model(pm::GenericPowerModel, gm::GenericGasModel, optimizer::JuMP.OptimizerFactory; solution_builder = get_solution)
-    termination_status, solve_time = optimize!(pm, gm, optimizer)
-    status = GasModels.parse_status(termination_status)
+function run_model(gdata::Dict{String,<:Any}, pdata::Dict{String,<:Any}, gtype::Type, ptype::Type, optimizer, build_method; gsp=[], psp=[], gext=[], pext=[], kwargs...)
```

Comment:
maybe the arguments `gsp` and `psp` would be more consistently named as `gm_solution_processors and `pm_solution_processors`?